File: Data Structures & Algorithms/house-robber-ii/submission-19.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) { 
        if (nums.length === 1) return nums[0]
        return Math.max(this.rec(0, nums, true, {}), this.rec(1, nums, false, {}))
    } 

    rec(i, nums, flag, memo) {  
        if (i in memo) return memo[i]
        if (i >= nums.length || (flag && i === nums.length - 1)) {
            return 0
        } 
        memo[i] = Math.max(
            nums[i] + this.rec(i + 2, nums, flag, memo),  
            this.rec(i + 1, nums, flag, memo))
        return memo[i]
    }
}

File: Data Structures & Algorithms/decode-ways/submission-19.py
class Solution:
    def numDecodings(self, s: str) -> int:
         
        def rec(i): 
            if i == len(s): 
                return 1 

            if s[i] == '0':
                return 0

            res = 0 
            res = rec(i + 1) 

            if i + 1 < len(s) and int(s[i:i+2]) <= 26: 
                res += rec(i + 2) 

            return res 

        return rec(0)